purchase_stock_picking_return_invoicing/models/purchase_order.py

```diff
@@ -0,0 +1,158 @@
+        return result
+
+    @api.multi
+    def action_view_invoice(self):
+        """Change super action for displaying only normal invoices."""
+        result = super(PurchaseOrder, self).action_view_invoice()
+        invoices = self.invoice_ids.filtered(
+            lambda x: x.type == 'in_invoice'
+        )
+        # choose the view_mode accordingly
+        if len(invoices) != 1:
+            result['domain'] = [('id', 'in', invoices.ids)]
+        elif len(invoices) == 1:
+            res = self.env.ref('account.invoice_supplier_form', False)
+            result['views'] = [(res and res.id or False, 'form')]
+            result['res_id'] = invoices.id
+        return result
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    qty_refunded = fields.Float(
+        compute="_compute_qty_refunded",
+        string='Refunded Qty',
+    )
+    qty_returned = fields.Float(
+        compute="_compute_qty_returned",
+        string='Returned* Qty',
+        help="This is ONLY the returned quantity that is refundable.",
+        store=True,
+    )
+
+    @api.depends('invoice_lines.invoice_id.state',
+                 'invoice_lines.quantity')
+    def _compute_qty_refunded(self):
+        for line in self:
+            inv_lines = line.invoice_lines.filtered(lambda x: (
+                (x.invoice_id.type == 'in_invoice' and x.quantity < 0.0) or
+                (x.invoice_id.type == 'in_refund' and x.quantity > 0.0)
+            ))
+            line.qty_refunded = sum(inv_lines.mapped(lambda x: (
+                x.uom_id._compute_quantity(x.quantity, line.product_uom)
+            )))
+
+    @api.depends(
+        'move_ids.state',
+        'move_ids.returned_move_ids.state',
+    )
+    def _compute_qty_returned(self):
+        """Made through read_group for not impacting in performance."""
+        ProductUom = self.env['product.uom']
+        groups = self.env['stock.move'].read_group(
+            [('purchase_line_id', 'in', self.ids),
+             ('state', '=', 'done'),
+             ('to_refund', '=', True),
+             ('location_id.usage', '!=', 'supplier')],
+            ['purchase_line_id', 'product_uom_qty', 'product_uom'],
+            ['purchase_line_id', 'product_uom'], lazy=False,
+        )
+        p = self._prefetch
+        # load all UoM records at once on first access
+        uom_ids = set(g['product_uom'][0] for g in groups)
+        ProductUom.browse(list(uom_ids), prefetch=p)
+        line_qtys = collections.defaultdict(lambda: 0)
+        for g in groups:
+            uom = ProductUom.browse(g["product_uom"][0], prefetch=p)
+            line = self.browse(g['purchase_line_id'][0], prefetch=p)
+            if uom == line.product_uom:
+                qty = g["product_uom_qty"]
+            else:
+                qty = uom._compute_quantity(
+                    g["product_uom_qty"], line.product_uom,
+                )
+            line_qtys[line.id] += qty
```

Comment:
Can you update line here?
`line['qty_returned'] += qty`

Reply:
Follow my comments from https://github.com/OCA/account-invoicing/pull/395#discussion_r198678688 about this topic

Reply:
@carlosdauden You saw the same point than me.

I have a similar case here:
https://github.com/odoo/odoo/commit/0f237f129e369b0edbab83209ce6bfd58436060d

The point of @pedrobaeza is that he don’t know the effects of use that way and he will need too much time to improvement a 3% of code and performance.

Then we shouldn’t stop this change that is a big performance improvement because of a smaller one.

You and I know that there is not others effects using this way then we can improvement it from other PR